### ISSUE TITLE
ENH: Integrating OpenBLAS Meson build

### ISSUE THREAD

**Issue by mtsokol** [scipy/scipy#21396]
### Is your feature request related to a problem? Please describe.

This issue is meant to discuss the feature and integration approach of Meson based OpenBLAS build.

There have been discussions around adding Meson based build to OpenBLAS (e.g., https://github.com/scipy/scipy/issues/12633#issuecomment-1078681133). In https://github.com/HaoZeke/OpenBLAS/tree/mesonBasic there's already a Meson port of OpenBLAS that supports `x86_64` (four subarchitectures) and `arm64` (only `armv8`). NumPy's and SciPy's test suites pass when this port is used.

### Describe the solution you'd like.

The initial PR is here: https://github.com/scipy/scipy/pull/21380. Rather than git submodule, it should be incorporated as a Meson submodule, and build from source with I think `meson submodule foreach ...` commands (?).

Other topics to discuss taken from https://github.com/scipy/scipy/pull/21380#issuecomment-2288784487:

- the desired behavior of the feature itself (I think this is a last fallback, after we've failed to locate not only another OpenBLAS but any other library like MKL etc.)
- the name of a `dev.py` flag
- what the CI job should look like (I think it has to use `pip install .`, not `dev.py`)

**Comment by andyfaff** [comment]
What does the meson submodule mean? Making everyone git clone/checkout openblas is quite heavy. Does a meson submodule only git clone or download openblas if it's asked to?

W.r.t a CI job, how long does it take to build?

My concern with the `dev.py` flag is that if we call it something like `--with-meson-openblas` people are going to get confused and think that's the default way to build with openblas.

**Comment by lucascolley** [comment]
The docs for meson subprojects are at https://mesonbuild.com/Subprojects.html

**Comment by nickodell** [comment]
>W.r.t a CI job, how long does it take to build?

Assuming it were built from source, I think around 10-40 minutes is the right ballpark.

There are some SciPy maintainers who maintain a pip package which contains binaries for OpenBLAS, and looking at their Actions logs, this takes about 10 to 30 minutes to build and test. https://github.com/MacPython/openblas-libs/actions

(There are workflows that take longer, but they're either build matrices that are building lots of different options, or doing something like emulating a different CPU architecture.)

@mattip might be a better person to ask, though.

**Comment by andyfaff** [comment]
What is the desired behaviour of the CI entry? Given that the build demands will be larger, it's important to justify that. Building OpenBLAS is already checked by the openblas-libs/build of the wheel, and most of the CI entries use OpenBLAS to run the scipy test suite. This means we've already checked most of that kind of functionality.

If the point of the test is to check that the dev.py flag works, then it may be less important to have such a test. Alternatively, we could hijack one of the existing CI entries to check that the flag works, but I think we'd definitely want to be using ccache (fcache?) to minimise the OpenBLAS build time.

**Comment by HaoZeke** [comment]
> What is the desired behaviour of the CI entry? Given that the build demands will be larger, it's important to justify that. Building OpenBLAS is already checked by the openblas-libs/build of the wheel, and most of the CI entries use OpenBLAS to run the scipy test suite. This means we've already checked most of that kind of functionality.

As I understand the CI job is meant to check that OpenBLAS can be built with `meson` as a fallback option when no existing version (including the wheel build, which doesn't use the `meson` build system) works. The build seems to complete with tests in around 25 minutes on CI so it shouldn't be too onerous. From a SciPy perspective then, is the wheel a better option than a `meson` subproject? I'd have thought not, since the `wheel` builds are "fatter" since they build symbols for a dynamic list of architectures instead of just the build system (currently the `meson` build doesn't support DYNAMIC_ARCH and uses the makefile native `getarch` system to determine the architecture and sub-architecture (e.g. HASWELL, NEHALEM etc).

> 
> If the point of the test is to check that the dev.py flag works, then it may be less important to have such a test. Alternatively, we could hijack one of the existing CI entries to check that the flag works, but I think we'd definitely want to be using ccache (fcache?) to minimise the OpenBLAS build time.

+1

**Comment by HaoZeke** [comment]
> What does the meson submodule mean? Making everyone git clone/checkout openblas is quite heavy. Does a meson submodule only git clone or download openblas if it's asked to?

Yes, as per the command line options in [the document](https://mesonbuild.com/Subprojects.html#commandline-options) @lucascolley linked to. Rephrasing the relevant bits, the subproject can be configured as an opt-in "absolute fallback" if all other detection fails. No major additional clones take place before that (the `.wrap` file defining the meson subproject is miniscule, but will need to be part of `scipy`). 

> W.r.t a CI job, how long does it take to build?
> 
> My concern with the `dev.py` flag is that if we call it something like `--with-meson-openblas` people are going to get confused and think that's the default way to build with openblas.

Maybe not even have a `dev.py` flag for it explicitly? With the subproject approach, `dev.py` can default to `--wrap-mode=nodownload` we only need to document that a user can, if the build fails due to not having `openblas` run `meson subprojects download` to get the source or `meson subprojects update` to update previously downloaded sources.

**Comment by mattip** [comment]
I am still stuck on the motivation of adding this to SciPy. I don't understand why SciPy should be building OpenBLAS from source.

> As I understand the CI job is meant to check that OpenBLAS can be built with meson as a fallback option when no existing version (including the wheel build, which doesn't use the meson build system) works. 

Doesn't the scipy wheel build use meson? I see meson is the [build system](https://github.com/scipy/scipy/blob/0866f813122dbe38a7e0c51ff534a9d78df24368/pyproject.toml#L18) in pyproject.toml.

Maybe I am missing something basic here.

**Comment by andyfaff** [comment]
I think the rationale is that some people have issues setting up BLAS. So this flag will take care of building + installing BLAS so it can be used as a fallback, and they don't need to deal with PKG_CONFIG_PATH, etc.

Whilst I can see the benefit of that, I don't think it deserves a special CI item, unless it's used as part of something else

**Comment by mattip** [comment]
> I think the rationale is that some people have issues setting up BLAS.

Thanks, that makes sense. Is there an issue open for that use case, or some other way to engage with such users? One of the justifications for the effort that went into making the scipy-openblas wheels was exactly this scenario. Perhaps these people are not aware of the scipy-openblas wheels or we can guide them through building the wheels in a way that will solve their use case, rather than offloading that problem from MacPython/openblas-libs onto SciPy.

**Comment by ev-br** [comment]
FWIW, I agree with Matti on this one. It's really not clear what is the benefit w.r.t either using a scipy-openblas wheel, or (if for some reason that does not cut it?) building OpenBLAS manually just before starting to build scipy and linking to this fresh build.

**Comment by rgommers** [comment]
Hey all, sorry for the lack of input here so far (this issue and work were my idea, but I got ill just when this discussion kicked off).

The tl;dr of my opinion is: 
1. We should do this, it can be a very useful tool for both end users and for debugging. 
2. The constraint is of course that it should be robust and performant enough that it solves a lot more problems than it creates. That speaks for itself - and doing this is reversible, so there isn't much risk here.
3. It should be opt-in only so users know what is happening, via a message saying _"Your build failed because a BLAS library could not be located. Please see http://scipy.github.io/devdocs/building/ for how to install one. Alternatively, you can try again with `pip install scipy -Csetup-args=-Dblas=openblas-src`, which will build OpenBLAS from source as part of the SciPy wheel build."_ (and we show the second half of that message only on platforms that we know should work)

Now to address the comments above on build time, difference with `scipy-openblas`, etc.

*Build time*: OpenBLAS builds in just under 50% of the time that SciPy itself builds (and with much lower memory requirements, because no C++). The reason that the `scipy-openblas` builds are so slow is that they have to be built with `DYNAMIC_ARCH=1`, meaning the same sources get rebuilt 5 times with different sets of SIMD flags. When one builds locally, the build can just target whatever the CPU supports, which is way faster (and slightly more optimal).

*Binary size*: for the same reason as the build time being faster, the binary size also ends up being smaller. It looks like they'll be about 20% - 25% smaller for x86-64.

*Build complexity*: copy-pasting a build command like `pip install scipy -Csetup-args=-Dblas=openblas-src` from an error message and it "just working" is way easier than having to install `scipy-openblas`, messing with `pkg-config`, adding a pre-loading snippet to `_distributor_init.py`, and then stringing everything together. Really one must then use `auditwheel` on it as well, because there is also the risk that `pip` will cache the built wheel, which is then going to break the next time the user installs `scipy` into a new venv (cause `scipy-openblas32` won't be present, nor is the dependency declared). The average maintainer still has trouble with this, and for users not familiar with SciPy it is much harder still - [this recent thread on Discourse](https://discuss.scientific-python.org/t/scipy-with-openblas/1531) illustrates that.

*What about `scipy-openblas`?* The `scipy-openblas` machinery and wheels are useful for nightlies and very much necessary for our own wheel builds, but it's not really usable as a debugging tool. The cycle of going through a large rebuild there (which so far needs Matt's shepherding), then uploading wheels to PyPI or anaconda.org, then downloading them locally and figuring out how to link against them has so far proven to enough overhead that it's not being used in practice. It is also very expensive CI time/jobs wise to rebuild and publish wheels for all platforms when you really just need a single OpenBLAS build that works on your machine.

So potential benefits:

- Helps users that now struggle with installing a BLAS library or messing with `delvewheel` & co. (especially on Windows this is a mess).
- Can make it a lot easier for ourselves to test changes in OpenBLAS (just build a wheel against a different commit). See [this comment](https://github.com/scipy/scipy/issues/21479#issuecomment-2496985081) for an example of how tricky it is right now to link in a different OpenBLAS build when debugging. Even experienced SciPy and scikit-learn maintainers tend to have a hard time here (we can probably count the number of people who can figure it out in less than 30 minutes on one hand).

  - The ability of figuring out easily whether a change in OpenBLAS affects some linear algebra related problem we're seeing is, in my opinion, in itself already valuable enough to implement this "build from source" option.

- (much more speculative, thanks to @mattip for the suggestion) could potentially be useful if we ever intend to unify thread pools. Right now we have a bunch of separate ones, with more opportunities for over-subscription.

* * *

A bit of context for how this can actually work. This was actually significantly more difficult to fully understand and do than I had anticipated. After writing the docs in https://github.com/mesonbuild/meson-python/pull/700 I have a pretty good idea about all the options for folding a shared library into a wheel build. The obvious alternative, building a static `libopenblas.a` and folding that into a regular shared library within the SciPy build config, unfortunately did not work. That static library approach is usually the most straightforward one, we use it for HiGHS and for vendored Fortran code for example, but OpenBLAS has internal machinery that seems to make this difficult. I seem to remember some summary about that being written somewhere, but can't find it anymore.

So, the way that seems to be working is this:
1. Users opts in and kicks off a wheel build with: `pip install scipy -Csetup-args=-Dblas=openblas-src` (or `uv`, or `python -m build`, etc.)
2. We have a small shim in `subprojects/openblas` that retrieves OpenBLAS sources and build config from the [Meson WrapDB service](https://mesonbuild.com/Wrap-dependency-system-manual.html)
3. A regular `libopenblas.so` gets built and marked for installation (we can name it as `libscipy_openblas.so` or whatever, to avoid clashes). It is built without `DYNAMIC_ARCH`, so each source file only gets compiled once, optimized for the CPU on the user's machine.
4. The rest of the SciPy build completes.
5. `meson-python` knows how to automatically fold in `libopenblas.so` into the wheel at `scipy.mesonpy.libs` and rewrite RPATHs with `patchelf` (on Windows slightly differently of course, since there are no rpaths).
